hs_core/tests/api/views/test_profile_required.py

```diff
@@ -0,0 +1,123 @@
+            superuser=False,
+            password=self.password,
+            groups=[]
+        )
+
+        self.user.userprofile.middle_name = 'Williams'
+        self.user.userprofile.state = 'Iowa'
+        self.user.userprofile.zipcode = 50112
+        self.user.userprofile.country = 'USA'
+        self.user.userprofile.date_of_birth = '06/21/1978'
+        self.user.userprofile.last_four_ss = 3476
+        self.user.userprofile.phone_1 = '641-234-4573'
+        self.user.userprofile.save()
+
+        self.client.login(username=self.username, password=self.password)
+
+    def tearDown(self):
+        if os.path.exists(self.temp_dir):
+            shutil.rmtree(self.temp_dir)
+        super(TestProfileRequiredViewFunctions, self).tearDown()
+
+    def test_view_create_resource_profile_complete(self):
+        # here we are testing the profile_required on the
+        # create_resource_select_resource_type view
+        # with our profile complete
+
+        url = reverse('create_resource_select_resource_type')
+        request = self.factory.get(url)
+        request.user = self.user
+
+        response = create_resource_select_resource_type(request)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_view_create_resource_profile_not_complete(self):
+        # here we are testing the profile_required decorator
+        # on a view when a user has not filled out their profile
+
+        # removing fields from profile
+        self.user.userprofile.last_four_ss=None
+        self.user.userprofile.phone_1=None
+        self.user.userprofile.save()
+
+
+        url = reverse('create_resource_select_resource_type')
+        request = self.factory.get(url)
+        request.user = self.user
+
+        response = create_resource_select_resource_type(request)
+
+        self.assertEqual(response.status_code, status.HTTP_302_FOUND)
+
+        # seeing if the redirect url is the one set in the
+        # profile_required decorator by default
+        redirect_url = response._headers['location'][1]
+        self.assertEqual(redirect_url, "/my-documents/#_")
+
+        my_documents_page, _ = RichTextPage.objects.get_or_create(title='My Documents', slug='my-documents')
+        response2 = self.client.get(redirect_url, follow=True)
+
+        # seeing if the redirect was successful and that the page
+        self.assertEqual(response2.status_code, status.HTTP_200_OK)
+        self.assertEqual(response2.context[-1]['_current_page'].id, my_documents_page.id)
+
+    def test_view_create_resource_profile_not_complete_with_referer(self):
+        # here we are testing the profile_required decorator
+        # on a view when a user has not filled out their profile
+        # and where the user is coming from a specific page
+
+        # removing fields from profile
+        self.user.userprofile.last_four_ss=None
+        self.user.userprofile.phone_1=None
+        self.user.userprofile.save()
```

Comment:
same comments for above function - actually I'm noticing that these tests are essentially identical I suggest you fold these two test cases into one test case using subtests (care of Dmitry: https://www.caktusgroup.com/blog/2017/05/29/subtests-are-best/). The first subtest would essentially be your first test case, then you could modify the request by adding the referer header and repeat the assertions in a second subtest (without having to repeat all the initial setup that is duplicated in both of these tests).

Reply:
doh, I just realized we're not on django 1.11 - but 1.8 - so scratch the 'subtest' suggestion - but the overall point still stands - one test would reduce duplication a lot here.